### PR TITLE
Let crd support strategic merge patch on metadata

### LINKED ISSUE #56348: CR is stuck with "orphan" finalizer
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

I have a CRD which is backed by a controller which creates and deletes other CRDs. The dependent CRDs have a controller owner reference set. If I delete the controller object with --cascade=true, everything works as expected and the resources are cleaned up. If I use --cascade=false, I see that an "orpan" finalizer is added. No owner references on one of the dependents is removed and the main object gets never deleted.

**What you expected to happen**:

I would have expected that the garbage collector removes the owner reference and then removes the "orphan" finalizer afterwards from the object, before finally deleting it.

Is that expected and the controller has to  remove the references and the "orphan" finalizer itself, or should the garbage collector do that? Do I have to set blockOrphanDeletion to false?

**How to reproduce it (as minimally and precisely as possible)**:

I delete it like this:
```
cluster/kubectl.sh delete mycrdcontroller mycrdinstance --cascade=false
```

The metadata of the controller object then looks like this:

```
apiVersion: kubevirt.io/v1alpha1
kind: VirtualMachineReplicaSet
metadata:
  creationTimestamp: 2017-11-24T12:44:54Z
  deletionGracePeriodSeconds: 0
  deletionTimestamp: 2017-11-24T12:45:02Z
  finalizers:
  - orphan
  name: testreplicaset2
  namespace: default
  resourceVersion: "24789"
  selfLink: /apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachinereplicasets/testreplicaset2
  uid: 4555ee8e-d115-11e7-9842-525400c2f26a

```

Owner references are still there on the dependants:

```
apiVersion: kubevirt.io/v1alpha1
  kind: VirtualMachine
  metadata:
    creationTimestamp: 2017-11-24T12:57:04Z
    generateName: test1
    labels:
      kubevirt.io/nodeName: node0
      myvm: myvm2
    name: test1vxl5h
    namespace: default
    ownerReferences:
    - apiVersion: v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: VirtualMachineReplicaSet
      name: testreplicaset2
      uid: 4555ee8e-d115-11e7-9842-525400c2f26a
```

and nothing happens.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #56348

cc @liggitt @nikhita @enisoc @mengqiy @ironcladlou 

The size of this fix is `+13/1-` (e6d7af1). Comparing it with converting the garbage collector to user JSON merge patch(https://github.com/kubernetes/kubernetes/pull/56595 `+125/−20`), I prefer this fix. I'll add unit tests if we can reach consensus.

The major criticism on this approach was that we shouldn't support strategic merge patch for CR partially. I'd like to defend this approach once more:
* From the perspective of the evolution of CRD, supporting strategic merge patch is a goal. And no matter how we achieve that goal, the schema of `ObjectMeta` should be defined by the APIServer, like what's done in this PR. So I regard this PR as a small step towards a larger goal, rather than a one-time hack.
* From the perspective of API consistency , `ObjectMeta` should be homogeneous across all resources stored on apiservers, including the supported operations on `ObjectMeta`.
* From users' perspective, they will receive a clear error message if they send a strategic merge patch for fields other than the `ObjectMeta`.

Another point raised by @ironcladlou was that we should "promote the use of one of the IETF PATCH standards" (https://github.com/kubernetes/kubernetes/issues/56348#issuecomment-347559931). My perspective is that the strategic merge patch is the only patch type that can safely patch list in a singe roundtrip, it should be supported by all resources. Currently only CRD doesn't support it.

```release-note
Custom resources now accept strategic merge patch against its `metadata` field. The patch will be rejected if it contains fields other than `metadata`.
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 1]
```diff
@@ -439,6 +439,9 @@ func (UnstructuredObjectConverter) Convert(in, out, context interface{}) error {
 func (UnstructuredObjectConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
 	if kind := in.GetObjectKind().GroupVersionKind(); !kind.Empty() {
 		gvk, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{kind})
+		if gvk.GroupVersion().Version == runtime.APIVersionInternal {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: cc @smarterclayton

**Comment by liggitt** [comment]
cc @kubernetes/sig-api-machinery-pr-reviews

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -220,7 +220,7 @@ KUBELET_HOST=${KUBELET_HOST:-"127.0.0.1"}
 # By default only allow CORS for requests on localhost
 API_CORS_ALLOWED_ORIGINS=${API_CORS_ALLOWED_ORIGINS:-/127.0.0.1(:[0-9]+)?$,/localhost(:[0-9]+)?$}
 KUBELET_PORT=${KUBELET_PORT:-10250}
-LOG_LEVEL=${LOG_LEVEL:-3}
+LOG_LEVEL=${LOG_LEVEL:-5}
```
Comment (sttts) [inline hack/local-up-cluster.sh]: debug change

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 3]
```diff
@@ -866,7 +871,11 @@ func StrategicMergeMapPatch(original, patch JSONMap, dataStruct interface{}) (JS
 	// each field in a strategic merge patch, we can't find the go struct tags. Hence, we can't easily  do a strategic merge
 	// for custom resources. So we should fail fast and return an error.
 	if _, ok := dataStruct.(*unstructured.Unstructured); ok {
-		return nil, mergepatch.ErrUnsupportedStrategicMergePatchFormat
+		schema, err = NewPatchMetaFromStruct(MetaOnly{})
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: What does this mean for patches again fields outside of meta? Will they apply like to a native type without any tags?

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: If yes, I would like to see a test case. If not, it must be rejected.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: only supporting a patch type on part of an object is bizarre to me

**Comment by liggitt** [comment]
> My perspective is that the strategic merge patch is the only patch type that can safely patch list in a singe roundtrip, it should be supported by all resources. Currently only CRD doesn't support it.

Is our position that all aggregated API servers must support strategic merge patch? I don't see a compelling reason to require that to participate in GC

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go [hunk 4]
```diff
@@ -866,7 +871,14 @@ func StrategicMergeMapPatch(original, patch JSONMap, dataStruct interface{}) (JS
 	// each field in a strategic merge patch, we can't find the go struct tags. Hence, we can't easily  do a strategic merge
 	// for custom resources. So we should fail fast and return an error.
 	if _, ok := dataStruct.(*unstructured.Unstructured); ok {
-		return nil, mergepatch.ErrUnsupportedStrategicMergePatchFormat
+		if _, ok := patch["metadata"]; !ok || len(patch) != 1 {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: @sttts I forgot to push this check. This will prevent strategic merge patch outside of "metadata".

**Comment by lavalamp** [comment]
cc @lavalamp

**Comment by lavalamp** [comment]
cc @jennybuckley

**Comment by lavalamp** [comment]
> Is our position that all aggregated API servers must support strategic merge patch?

Not thus far.

> I don't see a compelling reason to require that to participate in GC.

I tend to agree, it seems good to keep the bar low for participating.

**Comment by enisoc** [comment]
@sttts wrote:
> only supporting a patch type on part of an object is bizarre to me

In my opinion it's weird that we don't already support strategic merge patch on metadata inside CRs. We have all the tags we need via ObjectMeta, so why forbid it? As evidenced by this use case for GC, it's a useful thing to be able to do; I would say we just didn't get around to implementing it yet. Apparently the authors and reviewers of GC support for CRs also made the (IMO) reasonable assumption that patching metadata should work the same for any object that has ObjectMeta (which we enforce for everything made by CRD).

@liggitt wrote:
> Is our position that all aggregated API servers must support strategic merge patch?

We aren't necessarily taking that position by doing this. We're already in the state where GC doesn't work for anything that doesn't support SMP. Our goal right now is fixing GC specifically for CRs hosted through CRD. Broadening GC support to other things that don't support SMP is beyond the scope of this bugfix.

On a side note, I would expect SMP to come "free" for any aggregated API server that uses our libraries and uses compiled-in Go types. Is that not the case?

@liggitt wrote:
> What was the measured impact of simply doing a JSON merge patch on the expected field with the resourceVersion?

We haven't measured it yet because Chao's PR (#56595) shows there is already significant impact on the core architecture of the GC. Currently everything runs off of a dependency graph that only changes when object relationships change (not when arbitrary fields in those objects change). In order to use anything other than SMP, GC would need to know the latest state of the object fields so it can avoid clobbering other changes, which is a significant change in the algorithm that we would prefer not to cherry-pick to 1.8.x and 1.9.1.

@liggitt wrote:
> Experience with strategic merge patch from apply has not been good. Why do we want to expand it?

Strategic merge patch/diff is currently our best known solution for the problem of reconciling mutations made by automation (e.g. controllers, admission) with changes to the original source of truth (e.g. YAML in source control). Kubectl apply was an early proving ground for this and we expect to use SMP more in things like Declarative App Management. So for example, some parts of DAM might require strategic merge diff, which will require non-compiled-in CRs (such as those from CRD) to provide the SMP tags in some form (e.g. OpenAPI).

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 5]
```diff
@@ -439,6 +439,9 @@ func (UnstructuredObjectConverter) Convert(in, out, context interface{}) error {
...
+			return in, nil
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: What is the purpose of this?

**Comment by lavalamp** [comment]
Just to clarify, Chao's claim is that CRD is the only apiserver that would ever need this fix, assuming others use the generic apiserver library. But this is equivalent to mandating all apiservers support SMP on metadata. (Otherwise we have to fix GC too, which would render this change redundant.)

I agree it's weird to generally allow patching arbitrary subsets of objects, but allowing patching of metadata even if you don't understand the rest of the object seems very powerful. So I don't object to that.

So I really think it comes down to whether we want to mandate support for (at least metadata) SMP.

Currently I am inclined to say:

1. Let's take this patch.
2. Let's *not* set a precedent, and be 100% willing to reevaluate this if someone comes up with a valid use case.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go (hunk 5 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: This should be impossible - an internal unstructed is a coding error, and should never be allowed

**Comment by smarterclayton** [comment]
Given that strategic merge patch is unversioned this means we can never change the merge patch strategy on metadata ever again.  I’m not confident enough in SMR to say that is accurate.

I don’t think we should require all api servers support strategic merge patch without a lot more discussion - that statement above chills me to the bone...

**Comment by smarterclayton** [comment]
Also, what happens to an existing valid api server extension today that doesn’t support this?

**Comment by caesarxuchao** [comment]
> Also, what happens to an existing valid api server extension today that doesn’t support this?

Do we have apiservers that are not based on the generic apiserver library? If they don't support SMP, they can still use background GC. Foreground and orphan will stuck.

**Comment by caesarxuchao** [comment]
Can we agree on accepting this patch for 1.9? It's making CRD more powerful, and fixes 99% GC use cases. 

I can continue working on my other fix (#56595) that converts GC to use JSON merge patch, but that one requires more soak time.

**Comment by lavalamp** [comment]
> Given that strategic merge patch is unversioned this means we can never change the merge patch 

Ugh, this is seriously still true? We need to replace it with something more principled ASAP.

File: hack/local-up-cluster.sh [hunk 6]
```diff
@@ -115,16 +110,6 @@ if [ "${CLOUD_PROVIDER}" == "openstack" ]; then
     fi
 fi
 
-#set feature gates if using ipvs mode
```
Comment (lavalamp) [inline hack/local-up-cluster.sh]: Why are there so many changes to this file? It seems unrelated?

**Comment by lavalamp** [comment]
I think I still agree with what I wrote above:

1. Let's take this patch.
2. Let's *not* set a precedent, and be 100% willing to reevaluate this if someone comes up with a valid use case.

File: hack/local-up-cluster.sh (hunk 6 above)
Comment (caesarxuchao) [reply hack/local-up-cluster.sh]: Must be a rebase error. I'll finalize this PR.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 7]
```diff
@@ -439,6 +439,9 @@ func (UnstructuredObjectConverter) Convert(in, out, context interface{}) error {
 func (UnstructuredObjectConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
 	if kind := in.GetObjectKind().GroupVersionKind(); !kind.Empty() {
 		gvk, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{kind})
+		// if gvk.GroupVersion().Version == runtime.APIVersionInternal {
+		// 	return in, nil
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: @smarterclayton I wanted to treat the original version as the internal version of a custom resource. Otherwise we need special handling for `Unstructured` at https://github.com/kubernetes/kubernetes/pull/56606/files#diff-15bdfc59ec78ff09b509d4a347560a1bR244 and https://github.com/kubernetes/kubernetes/pull/56606/files#diff-15bdfc59ec78ff09b509d4a347560a1bR346.

**Comment by caesarxuchao** [comment]
@lavalamp @liggitt it's ready for review. PTAL. Thank you.

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go [hunk 8]
```diff
@@ -6746,3 +6747,100 @@ func TestUnknownField(t *testing.T) {
 		}
 	}
 }
+
+func TestStrategicMergePatchUnstructured(t *testing.T) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go]: When writing this test, I didn't notice we already had a unit test at `staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go`. Anyway, it doesn't hurt to have extra tests at different layers.

**Comment by lavalamp** [comment]
@liggitt So did I convince you we should take this at kubecon? :)

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go (hunk 7 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: Looks much better without these lines. But we should delete instead of comment out :)

**Comment by liggitt** [comment]
> > @liggitt: Is our position that all aggregated API servers must support strategic merge patch?
>
> @enisoc: We aren't necessarily taking that position by doing this.

We are by refusing to make GC work with a resource unless it supports SMP

> @lavalamp: So did I convince you we should take this at kubecon? :)

I still think the semantics of only supporting a patch type on part of an object are bizarre. Would openapi claim to accept strategic merge patch content-type for CRDs? That would lead clients like `kubectl apply` to compute SMP patches, rather than jsonmerge patches, and fail.

Also, the typecasts in patch handling smell particularly bad to me.

I'd still be more inclined to switch GC to a patch type that is more reasonable to support, rather than do this.

**Comment by sttts** [comment]
> I'd still be more inclined to switch GC to a patch type that is more reasonable to support, rather than do this.

I agree with @liggitt. I would rather see a fixed GC in 1.9.x with non-strategic patches than taking this shortcut.

**Comment by lavalamp** [comment]
>>> @liggitt: Is our position that all aggregated API servers must support strategic merge patch?

>> @enisoc: We aren't necessarily taking that position by doing this.

> @liggitt: We are by refusing to make GC work with a resource unless it supports SMP

I disagree. I don't think we are "refusing" permanently. This is just an expedient way to solve the only existing problem case.

File: staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go (hunk 4 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: is an empty patch ever allowed?

Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go]: @jennybuckley An empty strategic merge patch on custom resources is not allowed (It wasn't allowed before as well).
